### ISSUE TITLE
Vault Username-Password Credential does not work within Folder level

### ISSUE THREAD

**Issue by dendydo** [jenkinsci/hashicorp-vault-plugin#191]
I set up Jenkins integration with Hashicorp Vault via a plugin. Faced with the fact that when using the "Vault Username-Password Credential" plugin ignores Folder settings and uses only plugin global settings. If the global settings (/configure) of the plugin are not specified, then retrieving the secret through the "Vault Username-Password Credential" is impossible.

### Version report

Jenkins and plugins versions report: Jenkins: 2.289.2 
OS: Linux - 3.10.0-1160.11.1.el7.x86_64

---
ace-editor:1.1
active-directory:2.24
allure-jenkins-plugin:2.29.0
ansicolor:1.0.0
ant:1.11
antisamy-markup-formatter:2.1
apache-httpcomponents-client-4-api:4.5.13-1.0
authentication-tokens:1.4
basic-branch-build-strategies:1.3.2
blueocean:1.24.7
blueocean-autofavorite:1.2.4
blueocean-bitbucket-pipeline:1.24.7
blueocean-commons:1.24.7
blueocean-config:1.24.7
blueocean-core-js:1.24.7
blueocean-dashboard:1.24.7
blueocean-display-url:2.4.1
blueocean-events:1.24.7
blueocean-git-pipeline:1.24.7
blueocean-github-pipeline:1.24.7
blueocean-i18n:1.24.7
blueocean-jira:1.24.7
blueocean-jwt:1.24.7
blueocean-personalization:1.24.7
blueocean-pipeline-api-impl:1.24.7
blueocean-pipeline-editor:1.24.7
blueocean-pipeline-scm-api:1.24.7
blueocean-rest:1.24.7
blueocean-rest-impl:1.24.7
blueocean-web:1.24.7
bootstrap4-api:4.6.0-3
bootstrap5-api:5.0.1-2
bouncycastle-api:2.20
branch-api:2.6.4
build-token-trigger:1.0.0
build-user-vars-plugin:1.7
build-with-parameters:1.5.1
caffeine-api:2.9.1-23.v51c4e2c879c8
checks-api:1.7.0
cloudbees-bitbucket-branch-source:2.9.9
cloudbees-folder:6.15
command-launcher:1.6
conditional-buildstep:1.4.1
config-file-provider:3.8.0
copyartifact:1.46.1
credentials:2.5
credentials-binding:1.26
dashboard-view:2.17
display-url-api:2.3.5
docker-commons:1.17
docker-workflow:1.26
downstream-buildview:1.9
durable-task:1.37
echarts-api:5.1.2-2
email-ext:2.83
external-monitor-job:1.7
favorite:2.3.3
font-awesome-api:5.15.3-3
fortify:20.2.35
generic-webhook-trigger:1.74
git:4.7.2
git-client:3.7.2
git-parameter:0.9.13
git-server:1.9
github:1.33.1
github-api:1.123
github-branch-source:2.11.1
global-build-stats:1.5
handlebars:3.0.8
handy-uri-templates-2-api:2.1.8-1.0
hashicorp-vault-plugin:3.8.0
hp-application-automation-tools-plugin:6.9
htmlpublisher:1.25
influxdb:3.0.1
jackson2-api:2.12.3
javadoc:1.6
jdk-tool:1.5
jenkins-design-language:1.24.7
jira:3.5
jira-steps:1.6.0
jira-trigger:1.0.1
jjwt-api:0.11.2-9.c8b45b8bb173
job-restrictions:0.8
jquery:1.12.4-1
jquery-detached:1.2.1
jquery3-api:3.6.0-1
jsch:0.1.55.2
junit:1.51
kubernetes:1.30.0
kubernetes-client-api:5.4.1
kubernetes-credentials:0.9.0
ldap:2.7
leastload:3.0.0
lockable-resources:2.11
mailer:1.34
mask-passwords:3.0
matrix-auth:2.6.7
matrix-project:1.19
maven-plugin:3.12
mercurial:2.15
metrics:4.0.2.8
momentjs:1.1.1
monitoring:1.87.0
nexus-jenkins-plugin:3.11.20210716-075132.3b66565
nodelabelparameter:1.8.1
okhttp-api:3.14.9
pam-auth:1.6
parameterized-trigger:2.41
pipeline-build-step:2.13
pipeline-graph-analysis:1.11
pipeline-input-step:2.12
pipeline-milestone-step:1.3.2
pipeline-model-api:1.8.5
pipeline-model-declarative-agent:1.1.1
pipeline-model-definition:1.8.5
pipeline-model-extensions:1.8.5
pipeline-multibranch-defaults:2.1
pipeline-rest-api:2.19
pipeline-stage-step:2.5
pipeline-stage-tags-metadata:1.8.5
pipeline-stage-view:2.19
pipeline-utility-steps:2.8.0
plain-credentials:1.7
plugin-util-api:2.3.0
popper-api:1.16.1-2
popper2-api:2.5.4-2
prisma-cloud-jenkins-plugin:20.12.541
prometheus:2.0.10
pubsub-light:1.16
purge-job-history:1.6
rebuild:1.32
repository-connector:2.0.5
resource-disposer:0.16
role-strategy:3.1.1
run-condition:1.5
saml:2.0.7
scm-api:2.6.4
script-security:1.77
snakeyaml-api:1.29.1
sse-gateway:1.24
ssh-credentials:1.19
ssh-slaves:1.32.0
sshd:3.0.3
stashNotifier:1.20
structs:1.23
timestamper:1.13
token-macro:2.15
trilead-api:1.0.13
uno-choice:2.5.6
variant:1.4
windows-slaves:1.8
workflow-aggregator:2.6
workflow-api:2.46
workflow-basic-steps:2.23
workflow-cps:2.92
workflow-cps-global-lib:2.21
workflow-durable-task-step:2.39
workflow-job:2.41
workflow-multibranch:2.26
workflow-scm-step:2.13
workflow-step-api:2.23
workflow-support:3.8

### Reproduction steps

1. Install plugin HashiCorp Vault (https://plugins.jenkins.io/hashicorp-vault-plugin/)
2. Configure plugin in Jenkins in global settings (/configure): set Url and Credentials for auth in Vault (**vault root token**, but i try approle too)
3. Create in Jenkins credentials type "**Vault Username-Password Credential**": set path to secret, username and password key, engine version, credentialsId = **My-HvaultCred** and other. Press "Test Vault Secrets retrieval" button and it reports: "Successfully retrieved username ******** and the password"
4. Test this credentials in pipeline (**MyFolder**/MyPipeline) and it works ok:
`checkout([$class: 'GitSCM', branches: [[name: "master"]], doGenerateSubmoduleConfigurations: false, extensions: [[$class: 'CloneOption', depth: 1, noTags: true, shallow: false]], submoduleCfg: [], userRemoteConfigs: [[credentialsId: "**My-HvaultCred**", url: "${APP_REPO_URL}"]]])`
5. Clean plugin global settings (/configure)
6. Configure Folder settings **MyFolder** where pipeline is: set the same properties that i set in global (same url and same vault root token)
7. I try to check retrieval like in step 3, but it failed.
> FAILED to retrieve username key:
com.datapipe.jenkins.vault.exception.VaultPluginException: The credential id was not configured - please specify the credentials to use.
8. Start pipeline - it failed too.

> [Pipeline] checkout
[2021-09-07T15:24:13.565Z] The recommended git tool is: git
[2021-09-07T15:24:13.573Z] using credential *******-service-hvaultid-****-1
[2021-09-07T15:24:13.580Z] Cloning the remote Git repository
[2021-09-07T15:24:13.580Z] Avoid fetching tags
[Pipeline] }
[Pipeline] // stage
[Pipeline] stage
[Pipeline] { (Declarative: Post Actions)
[Pipeline] script
[Pipeline] {
[Pipeline] notifyBitbucket
[2021-09-07T15:24:13.714Z] found no commit info
[Pipeline] }
[Pipeline] // script
[Pipeline] script
[Pipeline] {
[Pipeline] deleteDir
[Pipeline] }
[Pipeline] // script
[Pipeline] }
[Pipeline] // stage
[Pipeline] }
[Pipeline] // timestamps
[Pipeline] }
[Pipeline] // withEnv
[Pipeline] }
[Pipeline] // node
[Pipeline] End of Pipeline
Also:   hudson.remoting.Channel$CallSiteStackTrace: Remote call to channel
		at hudson.remoting.Channel.attachCallSiteStackTrace(Channel.java:1800)
		at hudson.remoting.UserRequest$ExceptionResponse.retrieve(UserRequest.java:356)
		at hudson.remoting.Channel.call(Channel.java:1001)
		at com.datapipe.jenkins.vault.credentials.common.VaultHelper.getVaultSecret(VaultHelper.java:42)
		at com.datapipe.jenkins.vault.credentials.common.AbstractVaultBaseStandardCredentials.getVaultSecretKeyValue(AbstractVaultBaseStandardCredentials.java:75)
		at com.datapipe.jenkins.vault.credentials.common.VaultUsernamePasswordCredentialImpl.getUsername(VaultUsernamePasswordCredentialImpl.java:61)
		at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.createUsernameFile(CliGitAPIImpl.java:2390)
		at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandWithCredentials(CliGitAPIImpl.java:2043)
		at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.access$500(CliGitAPIImpl.java:86)
		at org.jenkinsci.plugins.gitclient.CliGitAPIImpl$1.execute(CliGitAPIImpl.java:608)
		at org.jenkinsci.plugins.gitclient.CliGitAPIImpl$2.execute(CliGitAPIImpl.java:837)
		at org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler$GitCommandMasterToSlaveCallable.call(RemoteGitImpl.java:160)
		at org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler$GitCommandMasterToSlaveCallable.call(RemoteGitImpl.java:153)
		at hudson.remoting.UserRequest.perform(UserRequest.java:211)
		at hudson.remoting.UserRequest.perform(UserRequest.java:54)
		at hudson.remoting.Request$2.run(Request.java:376)
		at hudson.remoting.InterceptingExecutorService.lambda$wrap$0(InterceptingExecutorService.java:78)
Also:   hudson.remoting.Channel$CallSiteStackTrace: Remote call to secret_managment_agent_1
		at hudson.remoting.Channel.attachCallSiteStackTrace(Channel.java:1800)
		at hudson.remoting.UserRequest$ExceptionResponse.retrieve(UserRequest.java:356)
		at hudson.remoting.Channel.call(Channel.java:1001)
		at org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler.execute(RemoteGitImpl.java:145)
		at sun.reflect.GeneratedMethodAccessor520.invoke(Unknown Source)
		at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
		at java.lang.reflect.Method.invoke(Method.java:498)
		at org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler.invoke(RemoteGitImpl.java:131)
		at com.sun.proxy.$Proxy116.execute(Unknown Source)
		at hudson.plugins.git.GitSCM.retrieveChanges(GitSCM.java:1223)
		at hudson.plugins.git.GitSCM.checkout(GitSCM.java:1301)
		at org.jenkinsci.plugins.workflow.steps.scm.SCMStep.checkout(SCMStep.java:129)
		at org.jenkinsci.plugins.workflow.steps.scm.SCMStep$StepExecutionImpl.run(SCMStep.java:97)
		at org.jenkinsci.plugins.workflow.steps.scm.SCMStep$StepExecutionImpl.run(SCMStep.java:84)
		at org.jenkinsci.plugins.workflow.steps.SynchronousNonBlockingStepExecution.lambda$start$0(SynchronousNonBlockingStepExecution.java:47)
		at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
com.datapipe.jenkins.vault.exception.VaultPluginException: The credential id was not configured - please specify the credentials to use.
	at com.datapipe.jenkins.vault.credentials.common.VaultHelper.retrieveVaultCredentials(VaultHelper.java:131)
	at com.datapipe.jenkins.vault.credentials.common.VaultHelper.access$100(VaultHelper.java:29)
	at com.datapipe.jenkins.vault.credentials.common.VaultHelper$SecretRetrieve.call(VaultHelper.java:104)
	at com.datapipe.jenkins.vault.credentials.common.VaultHelper$SecretRetrieve.call(VaultHelper.java:51)
	at hudson.remoting.UserRequest.perform(UserRequest.java:211)
	at hudson.remoting.UserRequest.perform(UserRequest.java:54)
	at hudson.remoting.Request$2.run(Request.java:376)
	at hudson.remoting.InterceptingExecutorService.lambda$wrap$0(InterceptingExecutorService.java:78)
	at org.jenkinsci.remoting.CallableDecorator.call(CallableDecorator.java:18)
	at hudson.remoting.CallableDecoratorList.lambda$applyDecorator$0(CallableDecoratorList.java:19)
	at jenkins.util.ContextResettingExecutorService$2.call(ContextResettingExecutorService.java:46)
	at jenkins.security.ImpersonatingExecutorService$2.call(ImpersonatingExecutorService.java:80)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Finished: FAILURE

### Results

Expected result:

The same behavior as withVault: if there are Folder settings, it uses them and no global setting needed. withVault works as expected.

Actual result:

"Vault Username-Password Credential" use only global Jenkins settings (/configure) and ignore Folder setting. It makes the given feature useless.

**Comment by dendydo** [comment]
I looked at the code and found that getting settings at the directory level is simply not included - there is no such code. There are sketches, but they do not seem to be used.

I cannot understand why the functionality is declared, if it does not exist and was not there.

**Comment by hchakrouun** [comment]
Hello,
Actually, the plugin is not respecting the desired fallback(pipeline than folder than... than Global) when looking for Vault configuration to use, It uses directly the global configuration: https://github.com/jenkinsci/hashicorp-vault-plugin/blob/master/src/main/java/com/datapipe/jenkins/vault/credentials/common/VaultHelper.java#L89.
I already fixed that for all credentials types, but I am running in another issue: When I create a **Vault certificate Credential**, the **keyStore key** field is not saved and i try to fix this too, and once this is done I will send a pull request to review my changes.
